### PR TITLE
SONARPY-1556 Parameters should properly support union type as declared type

### PR DESCRIPTION
**Pull request by joke1196**

### CODE REVIEW

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java
Comment (joke1196) [inline python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:1]: Sorry about this it is unrelated to the PR. I will just add some more coverage.

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java [hunk 1]
```diff
@@ -248,6 +250,28 @@ void test_none_annotation() {
...
+  @Test
+  void test_unknown_annotation() {
+    TypeAnnotation typeAnnotation = typeAnnotation(
+      "l : MyType"
+    );
+    assertThat(fromTypeshedTypeAnnotation(typeAnnotation)).isEqualTo(anyType());
+    InferredType declaredType = fromTypeAnnotation(typeAnnotation);
+    assertThat(declaredType).isInstanceOf(AnyType.class);
+  }
+
```
Comment (joke1196) [inline python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:274]: Sorry about this. This is not related to the PR. It just raises the coverage.

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java [hunk 2]
```diff
@@ -379,6 +459,10 @@ void test_type_from_protobuf() throws TextFormat.ParseException {
       "fully_qualified_name: \"mod.t\""
       ))
       .isEqualTo(InferredTypes.or(STR, INT));
+    assertThat(protobufType(
+      "kind: INSTANCE\n" +
+      "fully_qualified_name: \"typing._SpecialForm\""
+      )).isEqualTo(anyType());
```
Comment (joke1196) [inline python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:465]: Sorry about this as well. This is not related to the PR. It just raises the coverage.

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java [hunk 3]
```diff
@@ -282,6 +306,62 @@ void test_optional_type_annotations() {
...
+      "def f(x: int | A):",
+      "  x").type()).isEqualTo(anyType());
+
+    assertThat(lastExpression(
+      "def f(x: B | int):",
+      "  x").type()).isEqualTo(anyType());
+  }
+
+  @Test
+  void test_union_type_annotation_with_any_type() {
```
Comment (jdodinh) [inline python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:338]: Maybe the title of this test is a little bit misleading, because we are not testing a uniton type here.

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java [hunk 4]
```diff
@@ -282,6 +306,62 @@ void test_optional_type_annotations() {
...
+    assertThat(lastExpression(
+      "from typing import Any",
+      "def f(x: Any):",
+      "  x").type()).isEqualTo(anyType());
+  }
+
```
Comment (jdodinh) [inline python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:344]: The following case seems to fail:
```java
  @Test
  void test_declared_type_outside_of_method() {
    DeclaredType union = new DeclaredType(new SymbolImpl("Union", "typing.Union"), List.of(new DeclaredType(BuiltinTypes.INT),
      new DeclaredType(BuiltinTypes.STR)));
    assertThat(lastExpression(
      "class Foo():",
      "  x: int | str",
      "  x").type()).isEqualTo(union);
  }
```
I am not entirely sure that it is within the scope of this ticket, but maybe it would be worth investigating?

**Review by jdodinh — CHANGES_REQUESTED** [review]
LGTM, found a test case that fails, but not sure it needs to be addressed as part of this ticket.

File: python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java (hunk 4 above)
Comment (joke1196) [reply python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:344]: Ok so I have taken a look how I could test this. It is a bit complicated as we cannot use `lastExpression`. Also as discussed this yet out of scope so for now we can leave it as is and come back to it later on if needed.

Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/test/java/org/sonar/python/types/InferredTypesTest.java:345]: Probably the class attributes ticket would also be a blocker for this (although I'm not sure it would work in the current state of things once both tickets are merged).

A way to rewrite the test could make use of a separate function that instantiates the class and then accesses the attribute (similar to the tests in this [PR](https://github.com/SonarSource/sonar-python/pull/1653)) because currently the fact there is no function fails the test before it can actually test anything.

I think it makes sense to have such a test (either in this or in the other PR) to ensure we eventually cover this case though.

File: python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java [hunk 5]
```diff
@@ -235,9 +243,32 @@ private static DeclaredType declaredTypeFromTypeAnnotation(Expression expression
     if (expression.is(Kind.NONE)) {
       return new DeclaredType(builtinSymbols.get(BuiltinTypes.NONE_TYPE));
     }
+    if (expression.is(Kind.BITWISE_OR)) {
+      var binaryExpression = (BinaryExpression) expression;
+      var unionType = declaredUnionType(binaryExpression.leftOperand(), binaryExpression.rightOperand(), builtinSymbols);
+      if (unionType != null) {
+        return unionType;
+      }
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java:251]: You can probably just return unconditionally here, since the default is to return null.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
Looks good!
Small comment about formatting.
About the test that combines class attributes and union types, please let me know if you decide to add it, otherwise I can do it from my PR.

File: python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java [hunk 6]
```diff
@@ -19,6 +19,8 @@
  */
 package org.sonar.python.types;
 
+import static org.sonar.plugins.python.api.symbols.Symbol.Kind.CLASS;
```
[the remark this answers is not in the document]
Comment (joke1196) [reply python-frontend/src/main/java/org/sonar/python/types/InferredTypes.java]: Yes, I see I had to redo this already with the imports. I'll fix my configuration.
